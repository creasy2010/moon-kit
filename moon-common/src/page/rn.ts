import MoonCore from 'moon-core';

import * as fse from 'fs-extra';
import {join} from 'path';
import {IContext} from 'moon-core/declarations/typings/page';
import {IFileSaveOptions} from 'moon-core/declarations/typings/page';
import PageHooks from "./hook";
/**
 * @desc
 *
 * @使用场景
 *
 * @Date    2019/3/20
 **/
export async function genRnPage(context: IContext,hook?:PageHooks) {
  let {pageInfo, projectPath} = context;
  let prettiesConfig = {};
  try {
    prettiesConfig = await fse.readJSON(join(projectPath, 'pretties.json'));
  } catch (err) {}

  //把less , style内容记录下来,我砶修改
  await MoonCore.ReduxGen.buildPage({
    ...context,
    prettiesConfig,
    beforeSave: async (options: IFileSaveOptions, context: IContext) => {
      hook && hook.beforeSave.call(options, context);
      if (options.tplPath.endsWith('less.ejs')) {
        options.content = options.content.replace(
          '@import "~style/theme.less";',
          '',
        );
      }

      if (
        options.tplPath === 'index.tsx.ejs' ||
        options.tplPath === 'components/sub-components.tsx.ejs'
      ) {
        let keyName = 'index';
        if (options.param && options.param.subComp) {
          keyName = MoonCore.StringUtil.toLCamelize(
            options.param.subComp.fileName,
          );
          options.content = options.content
            .replace(`import './${options.param.subComp.fileName}.less';`, '')
            .replace(`className="${keyName}"`, '');
        } else {
          options.content = options.content
            .replace(`import './index.less';`, '')
            .replace(
              `className="${MoonCore.StringUtil.toLCamelize(
                context.pageInfo.pageKey,
              )}"`,
              '',
            );
        }

        options.content = options.content
          .replace(
            "import * as React from 'react';",
            `
          import React from 'react';
          import { StyleSheet, View , Text } from 'react-native';
          `,
          )
          // .replace("React.Component","Component")
          .replace(/<div/gi, '<View')
          .replace(/<\/div>/gi, '</View>');

        options.content = `${options.content}
          const styles = StyleSheet.create({
            ${keyName} : {
            }
          });
        `;
      }
      return options;
    },
    afterSave: async (options, context) => {
      hook && hook.afterSave.call(options, context);
    //   if(options.toSaveFilePath.includes("index.tsx")) {
    //
    //     let projectSrc  = projectPath;
    //     let pageKey = context.pageInfo.pageKey;
    //     let pageFilePath =join('pages', context.pageInfo.pagePath);
    //     for (let i = 0, iLen = pageInfo.actors.length; i < iLen; i++) {
    //       let actor = pageInfo.actors[i];
    //
    //       let reducerKey =  MoonCore.StringUtil.toLCamelize(pageKey+"-"+actor.fileName);
    //
    //       await MoonCore.CompileUtil.insertFile(join(projectSrc, 'src/redux/reducers/index.ts'), [
    //         {
    //           mark: '//mark1//',
    //           isBefore: true,
    //           content: `import ${reducerKey} from "@/${pageFilePath}/reducers/${actor.fileName}";`,
    //           check: (content): boolean => !content.includes(pageFilePath),
    //         },
    //         {
    //           mark: '//mark2//',
    //           isBefore: false,
    //           content: reducerKey+ ',',
    //           check: (content, rawContent): boolean =>
    //             !rawContent.includes(pageFilePath),
    //         },
    //       ]);
    //     }
    //
    //     await MoonCore.CompileUtil.insertFile(join(projectSrc, 'src/app.tsx'), [
    //       {
    //         mark: '//pagePath//',
    //         isBefore: true,
    //         content: `'${pageFilePath}/index',`,
    //         check: (content): boolean => !content.includes(pageFilePath),
    //       }
    //     ]);
    //   }
    },
  });
}
