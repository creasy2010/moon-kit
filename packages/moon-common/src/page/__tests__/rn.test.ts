/**
 * @desc
 *
 * @使用场景
 *
 * @Date    2019/7/25
 **/

import {tmpdir} from 'os';
import {join} from 'path';
import {remove} from  'fs-extra';
import MoonCore from  'moon-core';
import {genRnPage} from  '../rn';

describe('rn生成页面测试', () => {
  it('列表页面', async () => {
    let targetDir = join(tmpdir(), 'moon-temp/' ,Math.random()+"");
    await genRnPage({
      projectPath: targetDir,
      //@ts-ignore
      pageInfo: cashManPageDef,
    });

    //直接访问有问题..似乎会有 读取不到文件的情况发生;;
    let content = await MoonCore.JestUtil.readDirFiles(targetDir);
    expect(content).toMatchSnapshot();
    await remove(targetDir);
  });
});


let cashManPageDef =
  {
    "pagePath": "pages/test",
    pageKey: 'pagesTest',
    title: '测试',
    "actors": [
      {
        "fileName": "main",
        "datas": [{ "name": "request", "value": { "q": "", "pageNum": 0, "pageSize": 10 } }, { "name": "total", "value": 0 }, { "name": "list", "value": [], "schemaType": "fromValue" }],
        "events": [
          { "name": "modifyRequest", "comment": "修改查询条件数据", "content": "immerUtil.assign(draftState.request,payload);", "param": "" },
          { "name": "cleanList", "comment": "清空查询结果", "content": "draftState.list = [];", "param": "" },
          { "name": "queryResult", "comment": "", "content": "draftState.list = payload.list;\n        draftState.total = payload.total;\n        ", "param": "" }
        ]
      }
    ],
    "actions": [
      {
        "fileName": "action",
        "methods": [
          {
            "name": "modifySearch",
            "comment": "\n              普通条件查询可以走,commonChange\n              ",
            "content": "\n      dispatch({ type: Command.modifyRequest, payload: param});\n      //修改完直接查询;\n      if(options.isQuery) {\n        await this.query(options.isResetPage)\n      }\n      ",
            "param": "param,options:{\n      isQuery:boolean;\n      isResetPage:boolean;\n    }={isQuery:true,isResetPage:false}"
          },
          {
            "name": "nextPage",
            "comment": "查询下一页",
            "param": "",
            "content": "\n      let {request} = getData().main;\n      request.pageNum=request.pageNum+1;\n      dispatch({ type: Command.modifyRequest, payload: request });\n      await this.query();\n              "
          },
          {
            "name": "query",
            "comment": "以当前查询条件查询",
            "param": "isResetPage:boolean=false",
            "content": "\n            if(isResetPage) {\n              await dispatch({ type: Command.modifyRequest, payload: {pageNum:0} });\n            }\n          \n          let {request} =  getData().main;\n      //TODO 接口缺失 \n      // let {} = await api..XXapi();\n      \n      dispatch({ type: Command.queryResult, payload: {\n        total:0,\n        list:[]\n        } });\n        "
          }
        ]
      }
    ],
    "subComps": [{ "fileName": "info", "methods": [{ "name": "render", "comment": "", "content": "", "param": "" }] }],
    "lifeCycles": { "init": { "name": "init", "param": "", "content": "" }, "clean": { "name": "clean", "param": "", "content": "" } },
    "mainComp": { "methods": [] }
  };