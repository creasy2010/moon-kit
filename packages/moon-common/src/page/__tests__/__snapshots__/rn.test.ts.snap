// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rn生成页面测试 列表页面 1`] = `
Object {
  "src/pages/pages/test/actions/action.ts": "import {Command} from \\"../constant\\";
import {IAllReducerProps} from \\"../types\\";
import {redux} from  'moon-runtime';
import {PageModel} from './index';
import api from \\"@/api\\";

export default class \${action} extends redux.BaseAction<IAllReducerProps> {

  constructor(pageModel:PageModel) {
    super(pageModel);
  }

 
        /**
         * 
              普通条件查询可以走,commonChange
              
         */
        async modifySearch(param,options:{
      isQuery:boolean;
      isResetPage:boolean;
    }={isQuery:true,isResetPage:false}) {
            
      dispatch({ type: Command.modifyRequest, payload: param});
      //修改完直接查询;
      if(options.isQuery) {
        await this.query(options.isResetPage)
      }
      
         },
   
        /**
         * 查询下一页
         */
        async nextPage() {
            
      let {request} = getData().main;
      request.pageNum=request.pageNum+1;
      dispatch({ type: Command.modifyRequest, payload: request });
      await this.query();
              
         },
   
        /**
         * 以当前查询条件查询
         */
        async query(isResetPage:boolean=false) {
            
            if(isResetPage) {
              await dispatch({ type: Command.modifyRequest, payload: {pageNum:0} });
            }
          
          let {request} =  getData().main;
      //TODO 接口缺失 
      // let {} = await api..XXapi();
      
      dispatch({ type: Command.queryResult, payload: {
        total:0,
        list:[]
        } });
        
         },
   
}

//create by moon https://github.com/creasy2010/moon",
  "src/pages/pages/test/actions/index.ts": "import {Command} from '../constant';
import api from 'api';
import {redux} from 'moon-runtime';

import Action from './action';

import * as pagesTestMain from '../reducers/main';

import {IAllReducerProps} from '../types';

export class PageModel extends redux.BaseModel<IAllReducerProps> {
  models = {
    pagesTestMain,
  };

  actions = {
    action: new Action(this),
  };

  constructor() {
    super('PagesTest');
  }

  /**
   * 初始化数据
   */
  async init() {
    this.emit(Command.init, {
      //main  :{},
    });
  }
}

export default new PageModel();

//create by moon https://github.com/creasy2010/moon
",
  "src/pages/pages/test/components/info.less": "

.info {
}

",
  "src/pages/pages/test/components/info.tsx": "import React from 'react';
import {StyleSheet, View, Text} from 'react-native';

import * as T from '../types';

import pageModel from '../actions';
import {connect} from 'react-redux';
import {store2Props} from '../selectors';

type IInfoProps = T.IAllReducerProps & T.IInfoProps;

@connect<Partial<IInfoProps>, T.IInfoState>(
  store2Props,
  actions,
)
export default class Info extends React.Component<
  Partial<IInfoProps>,
  T.IInfoState
> {
  constructor(props: IInfoProps) {
    super(props);
  }

  /**
    
*/
  render() {
    let {main} = this.props;

    return (
      <View>
        <View />
      </View>
    );
  }
}

//create by moon https://github.com/creasy2010/moon

const styles = StyleSheet.create({
  info: {},
});
",
  "src/pages/pages/test/constant.ts": "export const pageKey = 'PagesTest';

export const BASE = pageKey + '_';

export enum Command {
  //通用修改数据方法
  commonChange = 'PagesTest_commonChange',
  init = 'PagesTest_INIT',
  clean = 'PagesTest_CLEAN',

  /*
        修改查询条件数据
    */
  main_modifyRequest = 'PagesTest_main_modifyRequest',

  /*
        清空查询结果
    */
  main_cleanList = 'PagesTest_main_cleanList',

  /*
        
    */
  main_queryResult = 'PagesTest_main_queryResult',
}

//create by moon https://github.com/creasy2010/moon
",
  "src/pages/pages/test/index.less": "

.pagesTest {
}


",
  "src/pages/pages/test/index.tsx": "import React from 'react';
import {StyleSheet, View, Text} from 'react-native';

import {connect} from 'react-redux';

import * as T from './types';
import pageModel from './actions';
import {store2Props} from './selectors';

import Info from './components/info';

@connect<Partial<T.IAllReducerProps>, any>(store2Props)
export default class PagesTest extends React.Component<
  Partial<T.IAllReducerProps>,
  any
> {
  componentDidMount() {
    pageModel.init();
  }

  componentWillUnmount() {
    pageModel.clean();
  }

  render() {
    let {main} = this.props;

    return (
      <View id=\\"pagesTest\\">
        <View />
      </View>
    );
  }
}

//create by moon https://github.com/creasy2010/moon

const styles = StyleSheet.create({
  index: {},
});
",
  "src/pages/pages/test/reducers/main.ts": "import {Command} from '../constant';
import _ from 'lodash';
import {IMainReducer} from '../types';
import produce from 'immer';

const INITIAL_STATE: IMainReducer = {
  isReady: false,

  request: {q: '', pageNum: 0, pageSize: 10},

  total: 0,

  list: [],
};

/*
修改查询条件数据
*/
export function modifyRequest(state: IMainReducer, payload: any): IMainReducer {
  return state;
}

/*
清空查询结果
*/
export function cleanList(state: IMainReducer, payload: any): IMainReducer {
  return state;
}

/*

*/
export function queryResult(state: IMainReducer, payload: any): IMainReducer {
  return state;
}

//create by moon https://github.com/creasy2010/moon
",
  "src/pages/pages/test/selectors.ts": "import {createSelector} from 'reselect';
import {IAllReducerProps} from './types';

export function store2Props({PagesTestMain}: any): IAllReducerProps {
  return {
    main: PagesTestMain,
  };
}

//衍生数据使用请参考:  https://github.com/reduxjs/reselect
//create by moon https://github.com/creasy2010/moon
",
  "src/pages/pages/test/types.ts": "import Actions from './actions';

export interface IMainReducer {
  isReady: boolean;
  isLoading?: boolean;

  request: IMainRequest;

  total: IMainTotal;

  list: IMainListSet;
}

export type IAllReducerProps = {
  main: IMainReducer;

  [name: string]: any;
};

export type IInfoProps = {};
export type IInfoState = {};

export interface IMainRequest {
  q?: string;
  pageNum?: number;
  pageSize?: number;
  [k: string]: any;
}
export type IMainTotal = number;
export type IMainListSet = IMainList[];

export interface IMainList {
  [k: string]: any;
}

//create by moon https://github.com/creasy2010/moon
",
}
`;
